### PR TITLE
switch: use new gfx

### PR DESCRIPTION
**Pull request by Cpasjuste**

### CODE REVIEW

File: src/video/switch/SDL_switchvideo.c [hunk 1]
```diff
@@ -257,6 +225,20 @@ SWITCH_CreateWindow(_THIS, SDL_Window *window)
         return SDL_SetError("Could not create NWindow from layer: 0x%x", rc);
     }
 
+    rc = nwindowSetDimensions(&wdata->nWindow, 1920, 1080);
```
Comment (carstene1ns) [inline src/video/switch/SDL_switchvideo.c:252]: I have not kept up with current development, so no idea if there is still this restriction:
Will this work in handheld mode (i.e. fixed size to 720p)?

Comment (fincs) [reply src/video/switch/SDL_switchvideo.c:252]: Yes. This is merely configuring the size of the framebuffer. The actual displayed region is configured using nwindowSetCrop.

File: src/video/switch/SDL_switchvideo.c [hunk 2]
```diff
@@ -164,72 +180,39 @@ SWITCH_GetDisplayModes(_THIS, SDL_VideoDisplay *display)
...
+        }
+        data = (SDL_WindowData *) SDL_GetFocusWindow()->driverdata;
     }
 
-    gfxConfigureResolution(mode->w + data->padding * 2, mode->h);
-    display->current_mode = *mode;
-    _this->windows->w = mode->w;
-    _this->windows->h = mode->h;
-    if (renderer) {
-        renderer->UpdateViewport(renderer);
```
Comment (fincs) [inline src/video/switch/SDL_switchvideo.c:232]: Shouldn't the viewport still be updated, or is that already done automatically by SDL somewhere else?

File: src/video/switch/SDL_switchvideo.c [hunk 3]
```diff
@@ -238,23 +221,52 @@ SWITCH_SetDisplayMode(_THIS, SDL_VideoDisplay *display, SDL_DisplayMode *mode)
...
+    }
+
+    rc = nwindowSetCrop(&wdata->nWindow, 0, 0, window->w, window->h);
+    if (R_FAILED(rc)) {
+        nwindowClose(&wdata->nWindow);
+        viCloseLayer(&wdata->viLayer);
+        return SDL_SetError("Could not set NWindow crop: 0x%x", rc);
+    }
+
+    wdata->egl_surface = SDL_EGL_CreateSurface(_this, (NativeWindowType) &wdata->nWindow);
```
Comment (fincs) [inline src/video/switch/SDL_switchvideo.c]: The cast to `(NativeWindowType)` isn't needed.

File: src/video/switch/SDL_switchvideo.c (hunk 2 above)
Comment (Cpasjuste) [reply src/video/switch/SDL_switchvideo.c:232]: Yes it's handled by SDL, with the defined function here: https://github.com/devkitPro/SDL/blob/switch-sdl2_gfx-new/src/video/switch/SDL_switchopengles.c#L48

File: src/video/switch/SDL_switchvideo.c (hunk 3 above)
Comment (Cpasjuste) [reply src/video/switch/SDL_switchvideo.c]: > Incompatible pointer types passing 'NWindow *' (aka 'struct NWindow *') to parameter of type 'NativeWindowType' (aka 'char *')

Comment (fincs) [reply src/video/switch/SDL_switchvideo.c]: Not in the next release of switch-mesa, which is _required_ for building and using this PR (see switch-refactor branch). It was changed to void*.

**Comment by rsn8887** [comment]
I might be missing something, but what is the logic of having a device with a 720p screen but only a 1080p SDL displaymode?
 
If I am in handheld mode, set the renderer to mode -1 (“desktop resolution”) and call SDL_GetCurrentDisplayMode then it should tell me that I have 1280x720 pixels, I think.

Are we supposed to change the window size and or use setlogicalsize to go between 720p and 1080p dynamically depending on the Switch dock state? And just not use Fullscreen Desktop anymore? 

Some more information:
Before this PR, I called CreateRenderer with mode -1 and FULLSCREENMODE_DESKTOP to use “current” and then SDL_GetCurrentDisplayMode(0, &DM) to check my resolution. SDL always returned 1280 by 720 regardless of docked or not, which was fine since a test by eye seeemed to indicate that even when docked it was running in that 720p scaled to 1080p mode. I could force 1080p by not using “-1” mode when opening the renderer, but then the viewport was completely wrong (shifted down).

It seems like this new code would fix the viewport issue, but now always returns 1080p display mode dimension regardless of docked state, which also seems wrong. Because the resolution is 720p in handheld mode and 720p could also be used in docked mode before. 

Or am I missing something?

Here’s the code that I think might break with this PR:
https://github.com/rsn8887/uae4all2/blob/2706d7e46124b629f20f4ac16389f440e2958f19/src/switch/sdl2_to_sdl1.cpp#L35

**Comment by Cpasjuste** [comment]
Hi !

I will need to take a closer look but from what I remember it now works like this :

If create window is called with fullscreen flag, you'll have to use SDL_SetDisplayMode to change the current mode (1920x1080 or 1280x720). If you create a window in windowed mode, you'll have to use SDL_SetWindowSize to change the window size. I'll take a closer look as soon as possible, but from all the testing I've done, I think it now correctly follow how SDL2 should works.

**Comment by rsn8887** [comment]
I see. I didn’t realize I could still use SetDisplayMode to change modes. I thought the only mode available in the new code would be 1080p!

So it should all work. I will test it once it is all merged and updated.

**Comment by Cpasjuste** [comment]
Would be better if you could test before merging :D. 

There is a minimal sample for testing here : https://github.com/Cpasjuste/switch-examples/tree/sdl2/sdl2, should be easier. Note that the 640x480 window size is for old SDL2.

**Comment by Cpasjuste** [comment]
I think you don't want fullscreen now, for easier "resolution change". Create a 1920x1080 window and use setWindowSize to change ... window size (viewport should be automatically handled). Else you'll have to handle viewport, or create a new window with the correct size after setVideoMode (or recreate a window without setVideMode I guess).

You may use this to check if you want 1080p? 
https://github.com/Cpasjuste/pplay/blob/master/pplay/main.cpp#L280

**Comment by rsn8887** [comment]
It wouldn't be a problem to use WindowSize, in fact it would be nice to easily use all 1080p mode pixels in docked mode for highest picture quality.

I don't think I can easily test this PR, because I would need a new mesa and libnx, too, judging from the previous comments?

**Comment by fincs** [comment]
Pinging @Cpasjuste. New version of libnx/switch-mesa is here already.

**Comment by Cpasjuste** [comment]
Hey @rsn8887 , were you able to test this PR with latest libnx/mesa etc?

**Comment by rsn8887** [comment]
No and I won't have time to anytime soon.

Just merge it and I will use it eventually to get 1080p native, like you suggested.

**Pull request merged**